src/Data/UTxO/Transaction/Cardano/Shelley.hs

```diff
@@ -0,0 +1,386 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{-# OPTIONS_HADDOCK prune #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-} -- needed for {#- HLINT ... #-}
+
+module Data.UTxO.Transaction.Cardano.Shelley
+    (
+    -- * Initialization
+      mkInit
+    , NetworkId (..)
+
+   -- * Constructing Primitives
+    , mkInput
+    , mkOutput
+    , mkShelleySignKey
+    , mkByronSignKey
+    , estimateFee
+
+    -- Internal
+    , Shelley
+    , minUTxOvalue
+    , minFeeA
+    , minFeeB
+    ) where
+
+
+import Cardano.Api.Typed
+    ( NetworkId, TxExtraContent (..), TxIn (..), TxOut (..) )
+import Cardano.Binary
+    ( serialize' )
+import Cardano.Crypto.Hash.Class
+    ( Hash (UnsafeHash), hashWith )
+import Cardano.Crypto.Signing
+    ( SigningKey (..) )
+import Cardano.Crypto.Wallet
+    ( xprv )
+import Cardano.Slotting.Slot
+    ( SlotNo (..) )
+import Control.Monad
+    ( replicateM )
+import Control.Monad.Fail
+    ( MonadFail )
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Short
+    ( toShort )
+import Data.Either.Extra
+    ( eitherToMaybe )
+import Data.UTxO.Transaction
+    ( ErrMkPayment (..), MkPayment (..) )
+import Data.UTxO.Transaction.Cardano.Helpers
+    ( ed25519ScalarMult )
+import Data.Word
+    ( Word32, Word64 )
+import GHC.Natural
+    ( naturalToInteger )
+import Numeric.Natural
+    ( Natural )
+
+import qualified Cardano.Api.Typed as Cardano
+import qualified Cardano.Chain.Common as Byron
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+import qualified Shelley.Spec.Ledger.Address.Bootstrap as Ledger
+import qualified Shelley.Spec.Ledger.Credential as Ledger
+
+-- | Construct a payment 'Init' for /Shelley/ from primitive types.
+--
+-- __examples__:
+--
+-- >>> mkInit Mainnet 430000
+-- >>> mkInit (Testnet (NetworkMagic 1234)) 430000
+--
+-- @since 2.0.0
+mkInit
+    :: NetworkId
+        -- ^ A network tag, Mainnet or Testnet with NetworkMagic specified
+    -> Word64
+        -- ^ A ttl expressed in slot number counted from the beginning of blockchain
+    -> Word64
+        -- ^ fee of tx as taken when constructing change outputs
+        -- see also estimateFee
```

Comment:
Hmmm. This last parameter looks wrong. Fee estimation should not happen at this level. This library is simply for constructing transactions. It assumes nothing about fee estimation.

Reply:
the reason is this https://github.com/input-output-hk/cardano-node/blob/master/cardano-api/src/Cardano/Api/Typed.hs#L1040

Reply:
TxFee determines TxBody Shelley and this is signed .... So I had several attempts to address fee (if you look at commit history)

Reply:
I see (and forgot about the explicit fee in the Shelley transactions for a moment). Okay, I'll look a bit more into this function which I sort of skip during the first review :+1: